src/key_info_managers/mod.rs

```diff
@@ -146,60 +146,82 @@ pub trait ManageKeyInfo {
     fn exists(&self, key_triple: &KeyTriple) -> Result<bool, String>;
 }
 
-/// Returns a Vec of the KeyInfo objects corresponding to the given application name and
-/// provider ID.
-///
-/// # Errors
-///
-/// Returns an error as a String if there was a problem accessing the Key Info Manager.
-pub fn list_keys(
-    manager: &dyn ManageKeyInfo,
-    app_name: &ApplicationName,
-    provider_id: ProviderID,
-) -> Result<Vec<parsec_interface::operations::list_keys::KeyInfo>, String> {
-    use parsec_interface::operations::list_keys::KeyInfo;
-
-    let mut keys: Vec<KeyInfo> = Vec::new();
-    let key_triples = manager.get_all(provider_id)?;
-
-    for key_triple in key_triples {
-        if key_triple.app_name() != app_name {
-            continue;
+// "+ Send + Sync" is needed for things like:
+// ```
+// let store_handle = self.key_info_store.read().expect("Key store lock poisoned");
+// let _ = store_handle.list_keys(&app_name, ProviderID::Pkcs11)?;
+// ```
+// to work because `store_handle` is a RWLockReadGuard of (dyn ManageKeyInfo + Send + Sync +
+// 'static). Did not work with only `impl dyn ManageKeyInfo` below. Maybe there is a way to
+// implement automagically the same methods on (dyn ManageKeyInfo + Send + Sync + 'static)
+// if they are implemented on dyn ManageKeyInfo but not sure.
+impl dyn ManageKeyInfo + Send + Sync {
+    /// Returns a Vec of the KeyInfo objects corresponding to the given application name and
+    /// provider ID.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error as a String if there was a problem accessing the Key Info Manager.
+    pub fn list_keys(
+        &self,
+        app_name: &ApplicationName,
+        provider_id: ProviderID,
+    ) -> Result<Vec<parsec_interface::operations::list_keys::KeyInfo>, String> {
+        use parsec_interface::operations::list_keys::KeyInfo;
+
+        let mut keys: Vec<KeyInfo> = Vec::new();
+        let key_triples = self.get_all(provider_id)?;
```

Comment:
Potentially dumb question, but wouldn't it make sense to extend `get_all` to take in 3 parameters: one `Option` for each of app name, provider and key name - and we filter everything in that one call, based on the combination of parameters provided. You're already iterating through all the keys in the KIM

Reply:
Yeah, and also because it seems like a handy way of dealing with arbitrary filters like this - just pass which combination you're interested in and you get it. (the name of the function isn't that specific either, it should be more like `get_all_per_provider`)

Reply:
>  (the name of the function isn't that specific either, it should be more like get_all_per_provider)

yeah agree with that one! I would say it's maybe better to change that in a subsequent PR? I can raise an issue for it

Reply:
cc #326